heuristic/delta_calculations.py

```diff
@@ -417,4 +417,127 @@ def regret_objective_function(state, employee, off_shifts, saturdays, days, L_C_
+"""
+
+def employee_shift_value(state, e, shift, saturdays, sundays, invalid_shifts, shift_combinations_violating_daily_rest, shift_sequences_violating_daily_rest, shifts_at_week, weeks, shifts_at_day, week, competency_score):
+    day = int(shift[0]/24)
+    daily_rest_error = regret_daily_rest_error(state, day, e, shift, invalid_shifts, shift_combinations_violating_daily_rest, shift_sequences_violating_daily_rest)
+    weekly_rest_error = regret_weekly_rest(state, shifts_at_week, e, week, shift)
+    partial_weekend_error = regret_partial_weekend(state, e, shifts_at_day, saturdays, sundays, day)
+    isolated_days_error = regret_isolated_days(state, e, shifts_at_day, day, weeks)
+    deviation_contracted_hours = regret_deviation_contracted_hours(state, e, shift, week, weeks)
+
+    return (
+            daily_rest_error 
+            + weekly_rest_error 
+            + partial_weekend_error 
+            + isolated_days_error 
+            + deviation_contracted_hours 
+            - 100 * competency_score
             )
+
+def regret_weekly_rest(state, shifts_at_week, e, week, shift):
+    actual_shifts = [(t, v) for t,v in shifts_at_week[week] if state.x[e,t,v] == 1 or (t,v) == (shift[0],shift[1])]
+    off_shift_periods = []
+    important = [7 * 24 * week, 7 * 24 * (week + 1) ]
+  
+    if len(actual_shifts) == 0:
+        off_shift_periods.append(float(important[1] - important[0]))
+        return min(72, off_shift_periods[0])
+
+    else:
+        if(actual_shifts[0][0] - important[0] >= 36):
+            off_shift_periods.append(actual_shifts[0][0] - important[0])
+            
+
+        if(important[1] - (actual_shifts[-1][0] + actual_shifts[-1][1]) >= 36):
+            off_shift_periods.append(important[1] - (actual_shifts[-1][0] + actual_shifts[-1][1]))
+
+        for i in range(len(actual_shifts) - 1):
+            if(actual_shifts[i+1][0] - (actual_shifts[i][0] + actual_shifts[i][1]) >= 36):
+                off_shift_periods.append(actual_shifts[i+1][0] - (actual_shifts[i][0] + actual_shifts[i][1]))
+
+    if len(off_shift_periods) != 0:
+        return min(72, max(off_shift_periods))
+    
+    return -200
+
+
+def regret_partial_weekend(state, e, shifts_at_day, saturdays, sundays, day):
+    if day in saturdays:
+        for t,v in shifts_at_day[day + 1]:
+            if(state.x.get((e, t, v))):
+                return -2
+            else:
+                return 0
+
+    elif day in sundays:
+        for t,v in shifts_at_day[day - 1]:
+            if(state.x.get((e, t, v))):
+                return -2
+            else:
+                return 0
+    else:
+        return 0
+
+
+def regret_isolated_days(state, e, shifts_at_day, day, weeks):
+    isolated_day = 0
+
+    if day in [0, 1, (len(weeks)*7 - 2), (len(weeks)*7 - 1)]:
+        return 0
+
+    
+    #Isolated Working Day
+    if sum(state.x[e,t,v] for t,v in (shifts_at_day[day - 1] + shifts_at_day[day + 1])) == 2:
```

Comment:
How can you be sure that this is an actual isolated working day? I cant see that you are asserting that `day` is a working day? I might have been missing something.

Also, I think we have different weights for isolated working and isolated off, so the function should rather return a tuple.

Reply:
Seems like there is an error here actually, so good catch. If the sum == 2 then it should not be an isolated working day. That means you are working all three days. Changed this to sum == 0 instead. If both the day before and the day after the shift in question is not worked, then it is an isolated working day

The last comment would not matter. Easy to only change the increment added to the isolated off days variable when finding such a day.